{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6da4648b",
   "metadata": {},
   "source": [
    "# scPrinter setup for Bos taurus\n",
    "Our goal? Being to characterize differential TF profiles between peak and late lactation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0272265",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6753ebf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import scprinter as scp\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pickle\n",
    "import torch\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "from scanpy.plotting.palettes import zeileis_28\n",
    "from tqdm.contrib.concurrent import *\n",
    "from tqdm.auto import *\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import json\n",
    "import csv\n",
    "import re\n",
    "from sklearn.preprocessing import OneHotEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b5b0f3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir=\"/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man\"\n",
    "genome_dir=\"/net/talisker/home/benos/mae117/.local/share/genomes/ARS-UCD2.0\"\n",
    "frags_dir_controls=\"/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/controls/fragments-str\"\n",
    "frags_dir_peaks=\"/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/peak/fragments-str\"\n",
    "frags_dir_late=\"/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/late/fragments-str\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a751ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "controls_files=os.listdir(frags_dir_controls)\n",
    "controls_files_frags=[i for i in controls_files if i.endswith(\".fragments.tsv.gz\")]\n",
    "controls_files_frags=sorted([os.path.join(frags_dir_controls, i) for i in controls_files_frags])\n",
    "controls_files_samples=[\"_\".join(re.split(\"[/_\\\\.]\", i)[11:13]) for i in controls_files_frags]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e8f346f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_files=os.listdir(frags_dir_peaks)\n",
    "peak_files_frags=[i for i in peak_files if i.endswith(\".fragments.tsv.gz\")]\n",
    "peak_files_frags=sorted([os.path.join(frags_dir_peaks, i) for i in peak_files_frags])\n",
    "peak_files_samples=[\"_\".join(re.split(\"[/_\\\\.]\", i)[11:13]) for i in peak_files_frags]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ab1454b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "late_files=os.listdir(frags_dir_late)\n",
    "late_files_frags=[i for i in late_files if i.endswith(\".fragments.tsv.gz\")]\n",
    "late_files_frags=sorted([os.path.join(frags_dir_late, i) for i in late_files_frags])\n",
    "late_files_samples=[\"_\".join(re.split(\"[/_\\\\.]\", i)[11:13]) for i in late_files_frags]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d063c363",
   "metadata": {},
   "source": [
    "## Create a custom ARS-UCD2.0 for scPrinter to work on!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "315618ae",
   "metadata": {},
   "source": [
    "### We need to build a [custom genome object](https://ruochiz.com/scprinter_doc/tutorials/custom_genome_tutorial.html) for cows RE scPrinter\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d95a5f",
   "metadata": {},
   "source": [
    "* Read in the chromosome sizes of the cow genome as a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8fb23f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#os.mkdir(os.path.join(work_dir, \"data/custom_genomes/ARS-UCD2.0\"))\n",
    "\n",
    "# sizes have to be read in like a dictionary\n",
    "chrom_sizes = {}\n",
    "#with open(os.path.join(genome_dir, \"ARS-UCD2.0.fa.sizes\")) as f:\n",
    "with open(os.path.join(genome_dir, \"bosTau9.chrom.sizes\")) as f:\n",
    "    for line in f:\n",
    "       (key, val) = line.split()\n",
    "       chrom_sizes[key] = int(val)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e6d9b95",
   "metadata": {},
   "source": [
    "* Create a bias file for ARDS-UCD2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96fcbaf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/net/talisker/home/benos/mae117/.cache/scprinter/Tn5_NN_model_py_v2.pt'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pretrain_Tn5_bias_model = scp.datasets.pretrained_Tn5_bias_model\n",
    "pretrain_Tn5_bias_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "73f15cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ee7991df1856470cbac8904f157d83c1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Predicting Tn5 bias:   0%|          | 0/1958 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scp.genome.predict_genome_tn5_bias(fa_file=str(os.path.join(genome_dir, \"ARS-UCD2.0.fa\")),\n",
    "                            save_name=f\"{work_dir}/data/bias.h5\",\n",
    "                            tn5_model=pretrain_Tn5_bias_model,\n",
    "                            context_radius=50,\n",
    "                            device=\"cuda\",\n",
    "                            batch_size=5000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be01585",
   "metadata": {},
   "source": [
    "* Create the custom genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "807005d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'test': ['chr1', 'chr12', 'chr24', 'chr3', 'chr6'],\n",
       "  'valid': ['chr20', 'chr8'],\n",
       "  'train': ['chr10',\n",
       "   'chr11',\n",
       "   'chr13',\n",
       "   'chr14',\n",
       "   'chr15',\n",
       "   'chr16',\n",
       "   'chr17',\n",
       "   'chr18',\n",
       "   'chr19',\n",
       "   'chr2',\n",
       "   'chr21',\n",
       "   'chr22',\n",
       "   'chr23',\n",
       "   'chr25',\n",
       "   'chr26',\n",
       "   'chr27',\n",
       "   'chr28',\n",
       "   'chr29',\n",
       "   'chr4',\n",
       "   'chr5',\n",
       "   'chr7',\n",
       "   'chr9',\n",
       "   'chrX',\n",
       "   'chrY']},\n",
       " {'test': ['chr16', 'chr2', 'chr21', 'chr27', 'chr9'],\n",
       "  'valid': ['chr12', 'chr17'],\n",
       "  'train': ['chr1',\n",
       "   'chr10',\n",
       "   'chr11',\n",
       "   'chr13',\n",
       "   'chr14',\n",
       "   'chr15',\n",
       "   'chr18',\n",
       "   'chr19',\n",
       "   'chr20',\n",
       "   'chr22',\n",
       "   'chr23',\n",
       "   'chr24',\n",
       "   'chr25',\n",
       "   'chr26',\n",
       "   'chr28',\n",
       "   'chr29',\n",
       "   'chr3',\n",
       "   'chr4',\n",
       "   'chr5',\n",
       "   'chr6',\n",
       "   'chr7',\n",
       "   'chr8',\n",
       "   'chrX',\n",
       "   'chrY']},\n",
       " {'test': ['chr11', 'chr15', 'chr19', 'chr29', 'chr4', 'chrY'],\n",
       "  'valid': ['chr22', 'chr7'],\n",
       "  'train': ['chr1',\n",
       "   'chr10',\n",
       "   'chr12',\n",
       "   'chr13',\n",
       "   'chr14',\n",
       "   'chr16',\n",
       "   'chr17',\n",
       "   'chr18',\n",
       "   'chr2',\n",
       "   'chr20',\n",
       "   'chr21',\n",
       "   'chr23',\n",
       "   'chr24',\n",
       "   'chr25',\n",
       "   'chr26',\n",
       "   'chr27',\n",
       "   'chr28',\n",
       "   'chr3',\n",
       "   'chr5',\n",
       "   'chr6',\n",
       "   'chr8',\n",
       "   'chr9',\n",
       "   'chrX']},\n",
       " {'test': ['chr10', 'chr14', 'chr18', 'chr20', 'chr26', 'chr5'],\n",
       "  'valid': ['chr21', 'chr6'],\n",
       "  'train': ['chr1',\n",
       "   'chr11',\n",
       "   'chr12',\n",
       "   'chr13',\n",
       "   'chr15',\n",
       "   'chr16',\n",
       "   'chr17',\n",
       "   'chr19',\n",
       "   'chr2',\n",
       "   'chr22',\n",
       "   'chr23',\n",
       "   'chr24',\n",
       "   'chr25',\n",
       "   'chr27',\n",
       "   'chr28',\n",
       "   'chr29',\n",
       "   'chr3',\n",
       "   'chr4',\n",
       "   'chr7',\n",
       "   'chr8',\n",
       "   'chr9',\n",
       "   'chrX',\n",
       "   'chrY']},\n",
       " {'test': ['chr13', 'chr17', 'chr23', 'chr25', 'chr7', 'chrX'],\n",
       "  'valid': ['chr10', 'chr18'],\n",
       "  'train': ['chr1',\n",
       "   'chr11',\n",
       "   'chr12',\n",
       "   'chr14',\n",
       "   'chr15',\n",
       "   'chr16',\n",
       "   'chr19',\n",
       "   'chr2',\n",
       "   'chr20',\n",
       "   'chr21',\n",
       "   'chr22',\n",
       "   'chr24',\n",
       "   'chr26',\n",
       "   'chr27',\n",
       "   'chr28',\n",
       "   'chr29',\n",
       "   'chr3',\n",
       "   'chr4',\n",
       "   'chr5',\n",
       "   'chr6',\n",
       "   'chr8',\n",
       "   'chr9',\n",
       "   'chrY']}]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def make_btau_splits(with_chr_prefix=True, include_mt=False):\n",
    "    \"\"\"\n",
    "    Create five 'reasonable' ARS-UCD2.0 chromosome splits (train/valid/test) in the\n",
    "    same structure as the human example. Partitions are disjoint and exhaustive.\n",
    "\n",
    "    Args:\n",
    "        with_chr_prefix (bool): if True, chromosomes are named 'chr1'...'chr29','chrX','chrY' (and 'chrMT' if included)\n",
    "                                if False, they are '1'...'29','X','Y' ('MT' if included)\n",
    "        include_mt (bool): include mitochondrial chromosome in the partitions.\n",
    "\n",
    "    Returns:\n",
    "        List[Dict[str, List[str]]]: five dicts with keys 'train', 'valid', 'test'\n",
    "    \"\"\"\n",
    "\n",
    "    # ARS-UCD2.0 autosomes 1..29 plus sex chromosomes\n",
    "    autosomes = [f\"chr{str(i)}\" for i in range(1, 30)]\n",
    "    sex = [\"chrX\", \"chrY\"]\n",
    "    mt = [\"MT\"] if include_mt else []\n",
    "    all_chroms = autosomes + sex + mt\n",
    "\n",
    "    # Helper to add/remove 'chr' prefix\n",
    "    def lab(c):\n",
    "        #return f\"chr{c}\" if with_chr_prefix else c\n",
    "        return f\"{c}\" if with_chr_prefix else c\n",
    "\n",
    "    # ----\n",
    "    # Pre-chosen, balanced-ish folds (mix large+medium+small autosomes, spread X/Y across folds)\n",
    "    # These are *chromosome labels without prefix*; we’ll map to prefixed labels at the end.\n",
    "    # Each fold lists its 'test' and 'valid' sets; 'train' is derived as set difference of all_chroms.\n",
    "\n",
    "    folds_spec = [\n",
    "        # Fold 1\n",
    "        {\n",
    "            \"test\":  [\"chr1\", \"chr3\", \"chr6\", \"chr12\", \"chr24\"],\n",
    "            \"valid\": [\"chr8\", \"chr20\"],\n",
    "        },\n",
    "        # Fold 2\n",
    "        {\n",
    "            \"test\":  [\"chr2\", \"chr9\", \"chr16\", \"chr21\", \"chr27\"],\n",
    "            \"valid\": [\"chr12\", \"chr17\"],\n",
    "        },\n",
    "        # Fold 3\n",
    "        {\n",
    "            \"test\":  [\"chr4\", \"chr11\", \"chr15\", \"chr19\", \"chr29\", \"chrY\"],\n",
    "            \"valid\": [\"chr7\", \"chr22\"],\n",
    "        },\n",
    "        # Fold 4\n",
    "        {\n",
    "            \"test\":  [\"chr5\", \"chr10\", \"chr14\", \"chr18\", \"chr20\", \"chr26\"],\n",
    "            \"valid\": [\"chr6\", \"chr21\"],\n",
    "        },\n",
    "        # Fold 5\n",
    "        {\n",
    "            \"test\":  [\"chr7\", \"chr13\", \"chr17\", \"chr23\", \"chr25\", \"chrX\"],\n",
    "            \"valid\": [\"chr10\", \"chr18\"],\n",
    "        },\n",
    "    ]\n",
    "\n",
    "    # Optionally distribute MT (if included), put it into validation of Fold 1\n",
    "    if include_mt:\n",
    "        folds_spec[0][\"valid\"] = folds_spec[0][\"valid\"] + [\"MT\"]\n",
    "\n",
    "    # Build full folds, enforcing disjoint/exhaustive partition per fold\n",
    "    folds = []\n",
    "    all_set = set(all_chroms)\n",
    "    for spec in folds_spec:\n",
    "        test = set(spec[\"test\"])\n",
    "        valid = set(spec[\"valid\"])\n",
    "        # Sanity: ensure declared sets exist\n",
    "        assert test.isdisjoint(valid), \"test and valid overlap in spec\"\n",
    "        assert test.issubset(all_set) and valid.issubset(all_set), \"spec includes unknown chromosomes\"\n",
    "        train = sorted(all_set - test - valid, key=lambda x: (x not in sex+mt, x))  # stable-ish sort\n",
    "\n",
    "        fold = {\n",
    "            \"test\":  sorted([lab(c) for c in test],  key=lambda x: (x not in {lab(\"X\"), lab(\"Y\"), lab(\"MT\")}, x)),\n",
    "            \"valid\": sorted([lab(c) for c in valid], key=lambda x: (x not in {lab(\"X\"), lab(\"Y\"), lab(\"MT\")}, x)),\n",
    "            \"train\": sorted([lab(c) for c in train], key=lambda x: (x not in {lab(\"X\"), lab(\"Y\"), lab(\"MT\")}, x)),\n",
    "        }\n",
    "        folds.append(fold)\n",
    "\n",
    "    return folds\n",
    "\n",
    "btau_splits = make_btau_splits(with_chr_prefix=True, include_mt=False)\n",
    "btau_splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "289052a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "momoo = scp.genome.Genome(\n",
    "    name=str(os.path.join(work_dir, \"data/custom_genomes/ARS-UCD2.0\")),\n",
    "    chrom_sizes=chrom_sizes,\n",
    "    gff_file=f\"{genome_dir}/ARS-UCD2.0.annotation.gtf\",\n",
    "    fa_file=f\"{genome_dir}/ARS-UCD2.0.fa\",\n",
    "    bias_file=f\"{work_dir}/data/bias.h5\",\n",
    "    blacklist_file=f\"{genome_dir}/ARS-UCD2.0.blacklist.bed\", # empty bed file\n",
    "    bg=None,\n",
    "    splits=btau_splits\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "309e55b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "pickle.dump(momoo, open(f\"{work_dir}/data/custom_genomes/ARS-UCD2.0\", \"wb\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ec2f30",
   "metadata": {},
   "source": [
    "## Create scPrinter object(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "018e7ebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<scprinter.genome.Genome at 0x7fa217871ad0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objects = []\n",
    "with (open(f\"{work_dir}/data/custom_genomes/ARS-UCD2.0\", \"rb\")) as openfile:\n",
    "    while True:\n",
    "        try:\n",
    "            objects.append(pickle.load(openfile))\n",
    "        except EOFError:\n",
    "            break\n",
    "\n",
    "momoo=objects[0]\n",
    "momoo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69956c50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/peak/fragments-str/SRR33155817.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/peak/fragments-str/SRR33155819.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/peak/fragments-str/SRR33155821.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/peak/fragments-str/SRR33155823.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/late/fragments-str/SRR33155816.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/late/fragments-str/SRR33155818.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/late/fragments-str/SRR33155820.fragments.tsv.gz',\n",
       " '/net/talisker/home/benos/mae117/Documents/careers/opalia/the-milk-man/data/experiments/late/fragments-str/SRR33155822.fragments.tsv.gz']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_frags=peak_files_frags+late_files_frags\n",
    "exp_frags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3f331e6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multiple fragments files detected, it is suggested to provide sample names to avoid barcode collision\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ada319c15f2e42348a69c3c8c0b30220",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Importing fragments:   0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bd1ab69f38b34b90b0cac71c93b4ccf4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start transferring insertions\n"
     ]
    }
   ],
   "source": [
    "printer_exp=scp.pp.import_fragments(\n",
    "    path_to_frags=exp_frags,\n",
    "    barcodes=[None] * len(exp_frags),\n",
    "    savename=os.path.join(work_dir, 'profiles/ars-ucd_exp_scprinter.h5ad'),\n",
    "    genome=momoo,\n",
    "    min_num_fragments=1000, min_tsse=7,\n",
    "    sorted_by_barcode=False,\n",
    "    low_memory=False,\n",
    "\n",
    "    # Added\n",
    "    auto_detect_shift=False,    # do NOT run the shape-sensitive auto-detect routine\n",
    "    plus_shift=4,               # canonical Tn5 offset for ATAC (forward strand)\n",
    "    minus_shift=-5,             # canonical Tn5 offset for ATAC (reverse strand),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "08ddc252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['experiments_peak_0',\n",
       " 'experiments_peak_1',\n",
       " 'experiments_peak_2',\n",
       " 'experiments_peak_3',\n",
       " 'experiments_late_0',\n",
       " 'experiments_late_1',\n",
       " 'experiments_late_2',\n",
       " 'experiments_late_3']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfs=[f+f\"_{i}\" for i, f in enumerate(peak_files_samples)]\n",
    "lfs=[f+f\"_{i}\" for i, f in enumerate(late_files_samples)]\n",
    "samples=pfs+lfs\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3f08c33d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SRR33155823', 'SRR33155816', 'SRR33155817', 'SRR33155821', 'SRR33155819', 'SRR33155818', 'SRR33155820', 'SRR33155822']\n"
     ]
    }
   ],
   "source": [
    "print (printer_exp.insertion_file.obs_names[:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b8e1d696",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['experiments_peak_0',\n",
       " 'experiments_peak_1',\n",
       " 'experiments_peak_2',\n",
       " 'experiments_peak_3',\n",
       " 'experiments_late_0',\n",
       " 'experiments_late_1',\n",
       " 'experiments_late_2',\n",
       " 'experiments_late_3']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "printer_exp.insertion_file.obs_names = samples\n",
    "printer_exp.insertion_file.obs_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba7ac58",
   "metadata": {},
   "source": [
    "## Call peaks on the fragments in a preset for training seq2PRINT model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5daefe08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import tempfile\n",
    "# import shutil\n",
    "# import subprocess\n",
    "\n",
    "# print(subprocess.check_output([\"macs2\", \"--version\"]).decode())\n",
    "\n",
    "# # Create a temporary bin dir and symlink macs2 → macs3\n",
    "# tmp_bin = tempfile.mkdtemp()\n",
    "# macs3_path = shutil.which(\"macs3\")\n",
    "# os.symlink(macs3_path, os.path.join(tmp_bin, \"macs2\"))\n",
    "\n",
    "# # Prepend to PATH so that scPrinter and subprocesses pick it up\n",
    "# os.environ[\"PATH\"] = f\"{tmp_bin}:{os.environ['PATH']}\"\n",
    "\n",
    "# print(subprocess.check_output([\"macs2\", \"--version\"]).decode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4f0ccb87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "continue\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/scprinter/preprocessing.py:842: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  peak = resize_bed_df(pd.read_csv(peak, sep=\"\\t\", header=None, comment=\"#\"), peak_width)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 500 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/talisker/home/benos/mae117/.conda/envs/tmm/lib/python3.11/site-packages/scprinter/peak.py:50: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  dt = pd.read_table(file, header=None)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering overlapping peaks based on peak summit score ..\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "No objects to concatenate",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[21]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Call peaks, this set of peaks are recommended to train seq2PRINT model\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43mscp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcall_peaks\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      3\u001b[39m \u001b[43m    \u001b[49m\u001b[43mprinter\u001b[49m\u001b[43m=\u001b[49m\u001b[43mprinter_exp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfrag_file\u001b[49m\u001b[43m=\u001b[49m\u001b[43mexp_frags\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcell_grouping\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# here we call peaks on the cells that are included in the final analyses\u001b[39;49;00m\n\u001b[32m      6\u001b[39m \u001b[43m    \u001b[49m\u001b[43mgroup_names\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mall\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpreset\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mseq2PRINT\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      8\u001b[39m \u001b[43m    \u001b[49m\u001b[43moverwrite\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\n\u001b[32m      9\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m     11\u001b[39m \u001b[38;5;66;03m# Fetched the cleaned peaks, save, it will be used in the next step\u001b[39;00m\n\u001b[32m     12\u001b[39m cleaned_peaks=pd.DataFrame(printer_exp.uns[\u001b[33m\"\u001b[39m\u001b[33mpeak_calling\u001b[39m\u001b[33m\"\u001b[39m][\u001b[33m'\u001b[39m\u001b[33mall_cleaned\u001b[39m\u001b[33m'\u001b[39m][:])\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/scprinter/preprocessing.py:858\u001b[39m, in \u001b[36mcall_peaks\u001b[39m\u001b[34m(printer, frag_file, cell_grouping, group_names, iterative_peak_merging, merge_across_groups, peak_width, clean_temp, preset, n_jobs, overwrite, sample_names, **kwargs)\u001b[39m\n\u001b[32m    856\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    857\u001b[39m         \u001b[38;5;28;01mfor\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m group_names:\n\u001b[32m--> \u001b[39m\u001b[32m858\u001b[39m             cleaned_peaks, _ = \u001b[43mclean_macs2\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    859\u001b[39m \u001b[43m                \u001b[49m\u001b[43m[\u001b[49m\u001b[43mname\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    860\u001b[39m \u001b[43m                \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m.\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprinter\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmacs2\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    861\u001b[39m \u001b[43m                \u001b[49m\u001b[43mpeak_width\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    862\u001b[39m \u001b[43m                \u001b[49m\u001b[43mprinter\u001b[49m\u001b[43m.\u001b[49m\u001b[43mgenome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    863\u001b[39m \u001b[43m                \u001b[49m\u001b[43mpreset\u001b[49m\u001b[43m=\u001b[49m\u001b[43mpreset\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    864\u001b[39m \u001b[43m                \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    865\u001b[39m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    866\u001b[39m             peak_calling[name + \u001b[33m\"\u001b[39m\u001b[33m_cleaned\u001b[39m\u001b[33m\"\u001b[39m] = cleaned_peaks\n\u001b[32m    867\u001b[39m printer.uns[\u001b[33m\"\u001b[39m\u001b[33mpeak_calling\u001b[39m\u001b[33m\"\u001b[39m] = peak_calling\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/scprinter/peak.py:391\u001b[39m, in \u001b[36mclean_macs2\u001b[39m\u001b[34m(name, outdir, peak_width, genome, filter_chr, blacklist_peak_width, n, fdr_threshold, use_quantile_scores, final_peak_width, max_iter, preset)\u001b[39m\n\u001b[32m    389\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(name) \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;28mlist\u001b[39m, np.ndarray]:\n\u001b[32m    390\u001b[39m     name = [name]\n\u001b[32m--> \u001b[39m\u001b[32m391\u001b[39m cleaned_peaks, summary = \u001b[43mmake_peaks_df_bioframe\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    392\u001b[39m \u001b[43m    \u001b[49m\u001b[43m[\u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m.\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn\u001b[49m\u001b[43m \u001b[49m\u001b[43m+\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m_summits.bed\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mn\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    393\u001b[39m \u001b[43m    \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    394\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpeak_width\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    395\u001b[39m \u001b[43m    \u001b[49m\u001b[43mgenome\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    396\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfilter_chr\u001b[49m\u001b[43m=\u001b[49m\u001b[43mfilter_chr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    397\u001b[39m \u001b[43m    \u001b[49m\u001b[43mblacklist_peak_width\u001b[49m\u001b[43m=\u001b[49m\u001b[43mblacklist_peak_width\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    398\u001b[39m \u001b[43m    \u001b[49m\u001b[43mn\u001b[49m\u001b[43m=\u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    399\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfdr_threshold\u001b[49m\u001b[43m=\u001b[49m\u001b[43mfdr_threshold\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    400\u001b[39m \u001b[43m    \u001b[49m\u001b[43muse_quantile_scores\u001b[49m\u001b[43m=\u001b[49m\u001b[43muse_quantile_scores\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    401\u001b[39m \u001b[43m    \u001b[49m\u001b[43mfinal_peak_width\u001b[49m\u001b[43m=\u001b[49m\u001b[43mfinal_peak_width\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    402\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmax_iter\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmax_iter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    403\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    404\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m cleaned_peaks, summary\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/scprinter/peak.py:272\u001b[39m, in \u001b[36mmake_peaks_df_bioframe\u001b[39m\u001b[34m(summit_files, names, peak_width, genome, filter_chr, blacklist_peak_width, n, fdr_threshold, use_quantile_scores, final_peak_width, max_iter)\u001b[39m\n\u001b[32m    270\u001b[39m         \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mUsually, it\u001b[39m\u001b[33m'\u001b[39m\u001b[33ms because you use the wrong genome name\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    271\u001b[39m         \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m272\u001b[39m peaks = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mconcat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpeaks_all\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    273\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mfinish clearing\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    274\u001b[39m \u001b[38;5;66;03m# Export the final result as a DataFrame; getting the top (or as many) n peaks based on the score and then resort based on genomic position.\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/pandas/core/reshape/concat.py:380\u001b[39m, in \u001b[36mconcat\u001b[39m\u001b[34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[39m\n\u001b[32m    377\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m copy \u001b[38;5;129;01mand\u001b[39;00m using_copy_on_write():\n\u001b[32m    378\u001b[39m     copy = \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m380\u001b[39m op = \u001b[43m_Concatenator\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    381\u001b[39m \u001b[43m    \u001b[49m\u001b[43mobjs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    382\u001b[39m \u001b[43m    \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    383\u001b[39m \u001b[43m    \u001b[49m\u001b[43mignore_index\u001b[49m\u001b[43m=\u001b[49m\u001b[43mignore_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    384\u001b[39m \u001b[43m    \u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m=\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    385\u001b[39m \u001b[43m    \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m=\u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    386\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlevels\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlevels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    387\u001b[39m \u001b[43m    \u001b[49m\u001b[43mnames\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnames\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    388\u001b[39m \u001b[43m    \u001b[49m\u001b[43mverify_integrity\u001b[49m\u001b[43m=\u001b[49m\u001b[43mverify_integrity\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    389\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    390\u001b[39m \u001b[43m    \u001b[49m\u001b[43msort\u001b[49m\u001b[43m=\u001b[49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    391\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    393\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m op.get_result()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/pandas/core/reshape/concat.py:443\u001b[39m, in \u001b[36m_Concatenator.__init__\u001b[39m\u001b[34m(self, objs, axis, join, keys, levels, names, ignore_index, verify_integrity, copy, sort)\u001b[39m\n\u001b[32m    440\u001b[39m \u001b[38;5;28mself\u001b[39m.verify_integrity = verify_integrity\n\u001b[32m    441\u001b[39m \u001b[38;5;28mself\u001b[39m.copy = copy\n\u001b[32m--> \u001b[39m\u001b[32m443\u001b[39m objs, keys = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_clean_keys_and_objs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobjs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    445\u001b[39m \u001b[38;5;66;03m# figure out what our result ndim is going to be\u001b[39;00m\n\u001b[32m    446\u001b[39m ndims = \u001b[38;5;28mself\u001b[39m._get_ndims(objs)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.conda/envs/tmm/lib/python3.11/site-packages/pandas/core/reshape/concat.py:505\u001b[39m, in \u001b[36m_Concatenator._clean_keys_and_objs\u001b[39m\u001b[34m(self, objs, keys)\u001b[39m\n\u001b[32m    502\u001b[39m     objs_list = \u001b[38;5;28mlist\u001b[39m(objs)\n\u001b[32m    504\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(objs_list) == \u001b[32m0\u001b[39m:\n\u001b[32m--> \u001b[39m\u001b[32m505\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mNo objects to concatenate\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    507\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m keys \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    508\u001b[39m     objs_list = \u001b[38;5;28mlist\u001b[39m(com.not_none(*objs_list))\n",
      "\u001b[31mValueError\u001b[39m: No objects to concatenate"
     ]
    }
   ],
   "source": [
    "# Call peaks, this set of peaks are recommended to train seq2PRINT model\n",
    "scp.pp.call_peaks(\n",
    "    printer=printer_exp,\n",
    "    frag_file=exp_frags,\n",
    "    cell_grouping=[None], # here we call peaks on the cells that are included in the final analyses\n",
    "    group_names=['all'],\n",
    "    preset='seq2PRINT',\n",
    "    overwrite=False\n",
    ")\n",
    "\n",
    "# Fetched the cleaned peaks, save, it will be used in the next step\n",
    "cleaned_peaks=pd.DataFrame(printer_exp.uns[\"peak_calling\"]['all_cleaned'][:])\n",
    "cleaned_peaks.to_csv(f'{work_dir}/peaks/seq2print_exp_narrowPeak.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f05861",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Call peaks using chromvar preset, this set of peak are recommended to be use as cell x peak for scATAC-seq data, or analysis\n",
    "scp.pp.call_peaks(\n",
    "    printer=printer_late,\n",
    "    frag_file=late_files_frags,\n",
    "    cell_grouping=[None], # here we call peaks on the cells that are included in the final analyses\n",
    "    group_names=['chromvar_all'],\n",
    "    preset='chromvar',\n",
    "    overwrite=False\n",
    ")\n",
    "\n",
    "# Fetched the cleaned peaks, save, it will be used in the next step\n",
    "cleaned_peaks = pd.DataFrame(printer_late.uns[\"peak_calling\"]['chromvar_late_cleaned'][:])\n",
    "cleaned_peaks.to_csv(f'{work_dir}/peaks/chromvar_late_regions.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "181b2336",
   "metadata": {},
   "source": [
    "* compare the two sets of peaks with the different presets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eb4ede8",
   "metadata": {},
   "outputs": [],
   "source": [
    "print (pd.DataFrame(printer_late.uns[\"peak_calling\"]['all_cleaned'][:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a51b83c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print (pd.DataFrame(printer_late.uns[\"peak_calling\"]['chromvar_all_cleaned'][:]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tmm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
